precise/iaf_psc_exp_ps_time_reversal.cpp

```diff
@@ -0,0 +1,720 @@
+    dhaene_quick1(0),
+    dhaene_quick2(0),
+    dhaene_tmax_lt_t1(0),
+    dhaene_max(0),
+    dhaene_det_spikes(0),
+    c0(0),
+    c1a(0),
+    c1b(0),
+    c3a(0),
+    c3b(0),
+    c4(0),
+    det_spikes(0)
+{}
+
+nest::iaf_psc_exp_ps_time_reversal::Buffers_::Buffers_(iaf_psc_exp_ps_time_reversal & n)
+  : logger_(n)
+{}
+
+nest::iaf_psc_exp_ps_time_reversal::Buffers_::Buffers_(const Buffers_ &, iaf_psc_exp_ps_time_reversal & n)
+  : logger_(n)
+{}
+
+/* ---------------------------------------------------------------- 
+ * Parameter and state extractions and manipulation functions
+ * ---------------------------------------------------------------- */
+
+
+//constants for time-reversal state space spike-detection algorithm
+
+void nest::iaf_psc_exp_ps_time_reversal::Parameters_::calc_const_spike_test_()
+{
+  //line corresponding to the final timestep i.e t_right: continuation
+  // of the curved boundary: a + I*b
+    a1_ =tau_m_ * tau_ex_;
+    a2_ =tau_m_ * (tau_m_ - tau_ex_);
+    a3_ =c_m_ * U_th_ * (tau_m_ - tau_ex_);
+    a4_ =c_m_ * (tau_m_ - tau_ex_);
+
+    //line joining endpoints of the envelope: \alpha I + \beta
+    b1_ =-tau_m_ * tau_m_;
+    b2_ =tau_m_ * tau_ex_;
+    b3_ =tau_m_*(tau_m_ - tau_ex_)- tau_m_*tau_m_ + tau_m_* tau_ex_;
+    b4_ =-tau_m_*tau_m_; //b1
+    b5_ =tau_m_*c_m_*U_th_;
+    b6_ =tau_m_*(tau_m_- tau_ex_);
+    b7_ =-c_m_*(tau_m_ - tau_ex_);
+
+    //envelope or curved boundary
+    c1_ =tau_m_/c_m_;
+    c2_ =(-tau_m_ * tau_ex_)/(c_m_*(tau_m_ - tau_ex_));
+    c3_ =(tau_m_ * tau_m_)/ (c_m_ * (tau_m_ - tau_ex_));
+    c4_ =tau_ex_/tau_m_;
+    c5_ =(c_m_ * U_th_)/tau_m_;
+    c6_ =1-(tau_ex_/tau_m_);
+
+    //parallel line
+
+    d1_ = tau_m_ *  c_m_ ;
+    d2_ = tau_m_ * tau_ex_;
+    d3_ = c_m_ * (tau_m_ - tau_ex_);
+}
+
+void nest::iaf_psc_exp_ps_time_reversal::Parameters_::get(DictionaryDatum & d) const
+{
+  def<double>(d, names::E_L, E_L_);
+  def<double>(d, names::I_e, I_e_);
+  def<double>(d, names::V_th, U_th_+E_L_);
+  def<double>(d, names::V_min, U_min_+E_L_);
+  def<double>(d, names::V_reset, U_reset_+E_L_);
+  def<double>(d, names::C_m, c_m_);
+  def<double>(d, names::tau_m, tau_m_);
+  def<double>(d, names::tau_syn_ex, tau_ex_);
+  def<double>(d, names::tau_syn_in, tau_in_);
+  def<double>(d, names::t_ref, t_ref_);
+
+}
+
+double nest::iaf_psc_exp_ps_time_reversal::Parameters_::set(const DictionaryDatum & d)
+{
+
+  updateValue<double>(d, names::tau_m, tau_m_);
+  updateValue<double>(d, names::tau_syn_ex, tau_ex_);
+  updateValue<double>(d, names::tau_syn_in, tau_in_);
+  updateValue<double>(d, names::C_m, c_m_);
+  updateValue<double>(d, names::t_ref, t_ref_);
+  updateValue<double>(d, names::I_e, I_e_);
+
+
+  // if U0_ is changed, we need to adjust all variables defined relative to U0_
+  const double ELold = E_L_;
+  updateValue<double>(d, names::E_L, E_L_);
+  const double delta_EL = E_L_ - ELold;
+
+  if(updateValue<double>(d, names::V_reset, U_reset_))
+    U_reset_ -= E_L_;
```

Comment:
According to our new coding standards, all single-line blocks need be enclosed in `{ }`.